### PR TITLE
JSON format for leadership schedule

### PR DESCRIPTION
**Pull request by newhoggy**

### CODE REVIEW

**Comment by newhoggy** [comment]
Resolves https://github.com/input-output-hk/cardano-node/issues/3678

**Comment by newhoggy** [comment]
Example output:

```
$ CARDANO_NODE_SOCKET_PATH=/private/var/folders/zh/ln41q4zs52x2fd61rxccmq640000gn/T/chairman/test-14a8914f9f284a51/socket/node-bft1 cardano-cli query leadership-schedule --testnet-magic 1925 --genesis /private/var/folders/zh/ln41q4zs52x2fd61rxccmq640000gn/T/chairman/test-14a8914f9f284a51/shelley/genesis.json --vrf-signing-key-file /private/var/folders/zh/ln41q4zs52x2fd61rxccmq640000gn/T/chairman/test-14a8914f9f284a51/node-pool1/shelley/vrf.skey --stake-pool-id $(cardano-cli stake-pool id --cold-verification-key-file /private/var/folders/zh/ln41q4zs52x2fd61rxccmq640000gn/T/chairman/test-14a8914f9f284a51/node-pool1/shelley/operator.vkey) --current
     SlotNo                          UTC Time
-------------------------------------------------------------
     185016                   2022-03-17 12:31:23.16 UTC
     185063                   2022-03-17 12:31:23.63 UTC
     185106                   2022-03-17 12:31:24.06 UTC
     185126                   2022-03-17 12:31:24.26 UTC
     185139                   2022-03-17 12:31:24.39 UTC
     185166                   2022-03-17 12:31:24.66 UTC
     185183                   2022-03-17 12:31:24.83 UTC
     185196                   2022-03-17 12:31:24.96 UTC
     185199                   2022-03-17 12:31:24.99 UTC
     185273                   2022-03-17 12:31:25.73 UTC
     185339                   2022-03-17 12:31:26.39 UTC
     185343                   2022-03-17 12:31:26.43 UTC
     185406                   2022-03-17 12:31:27.06 UTC
     185446                   2022-03-17 12:31:27.46 UTC
     185456                   2022-03-17 12:31:27.56 UTC
     185489                   2022-03-17 12:31:27.89 UTC
```

```
$ CARDANO_NODE_SOCKET_PATH=/private/var/folders/zh/ln41q4zs52x2fd61rxccmq640000gn/T/chairman/test-14a8914f9f284a51/socket/node-bft1 cardano-cli query leadership-schedule --testnet-magic 1925 --genesis /private/var/folders/zh/ln41q4zs52x2fd61rxccmq640000gn/T/chairman/test-14a8914f9f284a51/shelley/genesis.json --vrf-signing-key-file /private/var/folders/zh/ln41q4zs52x2fd61rxccmq640000gn/T/chairman/test-14a8914f9f284a51/node-pool1/shelley/vrf.skey --stake-pool-id $(cardano-cli stake-pool id --cold-verification-key-file /private/var/folders/zh/ln41q4zs52x2fd61rxccmq640000gn/T/chairman/test-14a8914f9f284a51/node-pool1/shelley/operator.vkey) --current --output-as json
[
    {
        "slotNumber": 185016,
        "slotTime": "2022-03-17T12:31:23.16Z"
    },
    {
        "slotNumber": 185063,
        "slotTime": "2022-03-17T12:31:23.63Z"
    },
    {
        "slotNumber": 185106,
        "slotTime": "2022-03-17T12:31:24.06Z"
    },
    {
        "slotNumber": 185126,
        "slotTime": "2022-03-17T12:31:24.26Z"
    },
    {
        "slotNumber": 185139,
        "slotTime": "2022-03-17T12:31:24.39Z"
    },
    {
        "slotNumber": 185166,
        "slotTime": "2022-03-17T12:31:24.66Z"
    },
    {
        "slotNumber": 185183,
        "slotTime": "2022-03-17T12:31:24.83Z"
    },
    {
        "slotNumber": 185196,
        "slotTime": "2022-03-17T12:31:24.96Z"
    },
    {
        "slotNumber": 185199,
        "slotTime": "2022-03-17T12:31:24.99Z"
    },
    {
        "slotNumber": 185273,
        "slotTime": "2022-03-17T12:31:25.73Z"
    },
    {
        "slotNumber": 185339,
        "slotTime": "2022-03-17T12:31:26.39Z"
    },
    {
        "slotNumber": 185343,
        "slotTime": "2022-03-17T12:31:26.43Z"
    },
    {
        "slotNumber": 185406,
        "slotTime": "2022-03-17T12:31:27.06Z"
    },
    {
        "slotNumber": 185446,
        "slotTime": "2022-03-17T12:31:27.46Z"
    },
    {
        "slotNumber": 185456,
        "slotTime": "2022-03-17T12:31:27.56Z"
    },
    {
        "slotNumber": 185489,
        "slotTime": "2022-03-17T12:31:27.89Z"
    }
]
```

The testnet for the above test was created via: https://github.com/input-output-hk/plutus-apps/pull/364

File: cardano-cli/src/Cardano/CLI/Output.hs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+module Cardano.CLI.Output
+  ( OutputAs(..)
+  ) where
+
+import Data.Eq
+import Text.Show
+
+data OutputAs = OutputAsJson | OutputAsText deriving (Eq, Show)
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Output.hs:8]: This should probably be in  `Cardano.CLI.Shelley.Output`. However I'm also wondering if this is necessary as we use `Maybe OutputFile` to determine if we are rendering as JSON to disk or if we are printing to stdout. Are there instances where we don't render as JSON to disk?

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]
Nice. A couple comments.

File: cardano-cli/src/Cardano/CLI/Output.hs (hunk 1 above)
Comment (newhoggy) [reply cardano-cli/src/Cardano/CLI/Output.hs:8]: I just had this idea that we could have `--json-output=FILE` and if users really want it in `stdout`, they can do this: `--json-output=/dev/stdout`.

Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/Output.hs:8]: Fair enough. How about modifying `OutputFile` to encompass this?

File: cardano-cli/src/Cardano/CLI/Shelley/Parsers.hs [hunk 2]
```diff
@@ -989,6 +991,22 @@ pQueryCmd =
...
+  <> Opt.metavar "OUTPUT_TYPE"
+  <> Opt.help "Output type.  One of 'json' or 'text'."
+  <> Opt.value OutputAsText
+  )
+  where
+    readerOutputAs :: ReadM OutputAs
+    readerOutputAs = do
+      v <- Opt.str @Text
+      case v of
+        "json" -> return OutputAsJson
```
[the remark this answers is not in the document]
Comment (newhoggy) [reply cardano-cli/src/Cardano/CLI/Shelley/Parsers.hs]: I went with your suggestion after all.

File: cardano-cli/src/Cardano/CLI/Output.hs (hunk 1 above)
Comment (newhoggy) [reply cardano-cli/src/Cardano/CLI/Output.hs:8]: I thought about this and I think it's better to follow the existing convention.

**Review by Jimbo4350 — APPROVED** [review]
LGTM! Just squash the commits :+1:

**Comment by newhoggy** [comment]
bors merge

**Pull request merged**